controllers/raycluster_controller.go

```diff
@@ -0,0 +1,327 @@
+	deleteOptions = client.DeleteOptions{PropagationPolicy: &deletePolicy}
+)
+
+//+kubebuilder:rbac:groups=ray.io,resources=rayclusters,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=ray.io,resources=rayclusters/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=ray.io,resources=rayclusters/finalizers,verbs=update
+//+kubebuilder:rbac:groups=route.openshift.io,resources=routes,verbs=patch;delete;get
+//+kubebuilder:rbac:groups=core,resources=secrets,verbs=get;create;patch;delete;get
+//+kubebuilder:rbac:groups=core,resources=services,verbs=patch;delete;get
+//+kubebuilder:rbac:groups=core,resources=serviceaccounts,verbs=patch;delete;get
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+// TODO(user): Modify the Reconcile function to compare the state specified by
+// the RayCluster object against the actual cluster state, and then
+// perform operations to make the cluster state reflect the state specified by
+// the user.
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.15.3/pkg/reconcile
+
+func (r *RayClusterReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	logger := ctrl.LoggerFrom(ctx)
+
+	var cluster rayv1.RayCluster
+
+	if err := r.Get(ctx, req.NamespacedName, &cluster); err != nil {
+		if !errors.IsNotFound(err) {
+			logger.Error(err, "Error getting RayCluster resource")
+		}
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	if cluster.ObjectMeta.DeletionTimestamp.IsZero() {
+		if !controllerutil.ContainsFinalizer(&cluster, CodeflareOAuthFinalizer) {
+			logger.Info("Add a finalizer", "finalizer", CodeflareOAuthFinalizer)
+			controllerutil.AddFinalizer(&cluster, CodeflareOAuthFinalizer)
+			if err := r.Update(ctx, &cluster); err != nil {
+				logger.Error(err, "Failed to update RayCluster with finalizer", logRequeueing, strTrue)
+				return ctrl.Result{RequeueAfter: requeueTime}, err
+			}
+		}
+	} else if controllerutil.ContainsFinalizer(&cluster, CodeflareOAuthFinalizer) {
+		err := r.deleteIfNotExist(
+			ctx, types.NamespacedName{Namespace: cluster.Namespace, Name: crbNameFromCluster(&cluster)}, &rbacv1.ClusterRoleBinding{},
+		)
+		if err != nil {
+			logger.Error(err, "Failed to remove OAuth ClusterRoleBinding.", logRequeueing, strTrue)
+			return ctrl.Result{RequeueAfter: requeueTime}, err
+		}
+		controllerutil.RemoveFinalizer(&cluster, CodeflareOAuthFinalizer)
+		if err := r.Update(ctx, &cluster); err != nil {
+			logger.Error(err, "Failed to remove finalizer from RayCluster", logRequeueing, strTrue)
+			return ctrl.Result{RequeueAfter: requeueTime}, err
+		}
+		logger.Info("Successfully removed finalizer.", logRequeueing, strFalse)
+		return ctrl.Result{}, nil
+	}
+
+	if val, ok := cluster.ObjectMeta.Annotations["codeflare.dev/oauth"]; !ok || val != "true" {
```

Comment:
I'll turn the string to lowercase here

Reply:
To uppercase would be best as the annotation in the RayCluster when created ends up being `codeflare.dev/oauth: 'True'`

Reply:
Sorry. I meant for checking i.e. val.Lower = "true"

Reply:
I ended up finding this: https://pkg.go.dev/strconv#ParseBool which will be better in the long run since it allows for any truthy value of the annotation to be interpreted correctly. If the value isn't interpretable by the package  (i.e. returns an error) the default behavior is to treat it as false

Reply:
That is actually pretty cool, nice one Kevin!